File: 202507July/070725_Solution.java
class Solution {
    public int maxEvents(int[][] events) {
        Arrays.sort(events, (a, b) -> a[0] - b[0]);

        int lastDay = 0;
        for (int[] event: events) {
            lastDay = Math.max(lastDay, event[1]);
        }

        PriorityQueue<Integer> pq = new PriorityQueue<>();

        int count = 0;
        int i = 0;
        for (int d = 1; d <= lastDay; ++d) {
            while (!pq.isEmpty() && pq.peek() < d) {
                pq.poll();
            }

            while (i < events.length && events[i][0] <= d) {
                pq.add(events[i][1]);
                i++;
            }

            if (!pq.isEmpty()) {
                pq.poll();
                count++;
            }
        }

        return count;
    }
}

File: 202507July/070725/070725_solution.cs
public class Solution {
    public int MaxEvents(int[][] events) {
        Console.WriteLine("abc");
        var ans = 0;

        Array.Sort(events, (a, b) => a[0] - b[0]);
        var lastDay = 0;
        for(var i = 0; i < events.Length; i++) {
            lastDay = Math.Max(lastDay, events[i][1]);
        }

        var heap = new PriorityQueue<int, int>();
        var nextAvailableEventIndex = 0;
        for(var i = 1; i <= lastDay; i++) {
            while(nextAvailableEventIndex < events.Length && events[nextAvailableEventIndex][0] == i) {
                heap.Enqueue(events[nextAvailableEventIndex][1], events[nextAvailableEventIndex][1]);
                nextAvailableEventIndex++;
            }
            
            while(heap.Count > 0 && heap.Peek() < i) {
                heap.Dequeue();
            }

            if(heap.Count > 0) {
                heap.Dequeue();
                ans++;
            }
        }


        return ans;
    }
}
